### ISSUE TITLE
Check variance on pre-3.12 `TypeVar`s

### ISSUE THREAD

**Issue by decorator-factory** [DetachHead/basedpyright#1615]
### Description

See:
https://github.com/DetachHead/basedpyright/issues/763
https://github.com/DetachHead/basedpyright/issues/395
https://github.com/python/mypy/issues/735
https://discuss.python.org/t/why-do-type-checkers-not-warn-when-incorrect-variance-is-used/104668

Type checkers don't warn when incorrect variance is used with old-style `TypeVar`s.

```py
from dataclasses import dataclass
from typing import TypeVar, Generic

T_co = TypeVar("T_co", covariant=True)

@dataclass
class OldDataclass(Generic[T_co]):
    x: T_co  # error expected, not reported
    y: list[T_co]  # error expected, not reported

class OldBox(Generic[T_co]):
    def __init__(self, value: T_co) -> None:
        self.value: T_co = value  # error expected, not reported

class OldShelf(Generic[T_co]):
    def __init__(self, values: list[T_co]) -> None:
        self._values: list[T_co] = values

    def get_values(self) -> list[T_co]:  # error expected, not reported
        return self._values

    def set_values(self, new_values: list[T_co]) -> None:  # error expected, not reported
        self._values = new_values
```

From the discourse thread, pyright copied this bug from mypy. I think it's going to cause pain for people when they migrate from `TypeVar` to the modern syntax and only then realize that they used the wrong variance for their generic class all along. And, of course, it's just unsound.
I propose two alternate solutions:

1. Add an optional diagnostic rule for `covariant=` and `contravariant=` parameters on `TypeVar`. If you need to support Python 3.10 and 3.11 (or 3.12 and you want `default`), use `typing_extensions.TypeVar` with `infer_variance=True` instead. This should be pretty easy to implement.
2. Actually validate the variance. Just pretend `infer_variance` is on instead of `covariant=` or `contravariant=`, and show an error if the inferred variance doesn't match the stated variance. Not sure how difficult it would be to implement. The error will be confusing because it doesn't say why the variance is wrong. But I guess that's not any worse than new-style type variables that don't tell you why they're invariant.

**Comment by decorator-factory** [comment]
Nitpicking note: the error message for the rare case when parameter types are checked is misleading
```py
class OldBox(Generic[T_co]):
    def set_value(self, value: T_co) -> None:
        #                      ^^^^
        # Covariant type variable cannot be used in parameter type
```
Covariant type variables _can_ be used _somewhere in the parameter type_ (e.g. `param: Callable[[T_co], None]`), but they cannot be used _as the parameter type_ (or in a covariant position in a different way).

**Comment by DetachHead** [comment]
pyright does have a check for this, but only in the most basic scenario (where the generic is the only part of the type annotation):

```py
from typing import Generic, TypeVar


T_co = TypeVar("T_co", covariant=True)

class Foo(Generic[T_co]):
    def foo(self, value: T_co): ... # error
    def bar(self, value: T_co | None): ... # no error
```

i think this is the same issue as #764
